In [0]:
import numpy as np # 1
import pandas as pd # 2
import matplotlib.pyplot as plt # 3
from sklearn.cluster import KMeans # 4

Primeiramente vamos importar o dataset

Não foi utilizazo o dataset completo utilizei apenas 4 variaveis para fazer o agrupamento

'ESPVIDA', 'AGREGA', 'IDHM', 'POP', 'NOME_AGREGA'



In [0]:
headers = ['ESPVIDA', 'AGREGA', 'IDHM', 'POP', 'NOME_AGREGA']
dataset = pd.read_excel("./data_IDH.xlsx", encoding = "ISO-8859-1", decimal=",", header=None, names=headers)

Para realização do agrupamento de dados foi utilizado a biblioteca Kmeans.

Antes de utiliza-la tratamos as variaveis de predição forçando-as a ser do tipo float para garantir a estruturação

abrimos a variavel kmeans para receber o objeto de sua biblioteca e definimos o numero de agrupamentos para 4

'ESPVIDA', 'AGREGA', 'IDHM', 'POP'


In [0]:
for col in  dataset.columns[0:4]:
    dataset[col] = dataset[col].astype(float)

X = dataset.iloc[:, 0:4]
y = dataset.iloc[:, 4]

kmeans = KMeans(n_clusters=4, random_state=0).fit(X)
X_clustered = kmeans.fit_predict(X)


Agora gerimos o resultudado esperando o retorno de uma tabela com resultado de 'ClusterNumber' ao qual seria o agrupamento que 'NOME_AGREGA' ou onde os dados daquela região provavelmente estariam inserido. 

Como resultado tivemos um agrupamento de 0 a 3 ou seja 4 agrupamentos distintos.

In [0]:

results = dataset[['NOME_AGREGA']].copy()
results['clusterNumber'] = X_clustered
results


NOME_AGREGA  clusterNumber
0               BRASIL              1
1               BRASIL              1
2               BRASIL              1
3               BRASIL              1
4               BRASIL              1
5                   RO              2
6                   RO              2
7                   RO              2
8                   RO              2
9                   RO              2
10                  AC              2
11                  AC              2
12                  AC              2
13                  AC              2
14                  AC              2
15                  AM              2
16                  AM              2
17                  AM              2
18                  AM              2
19                  AM              2
20                  RR              2
21                  RR              2
22                  RR              2
23                  RR              2
24                  RR              2
25                  PA              2
26                  PA              2
27                  PA              2
28                  PA              2
29                  PA              2
..                 ...            ...
160  RM Belo Horizonte              2
161  RM Belo Horizonte              2
162  RM Belo Horizonte              2
163  RM Belo Horizonte              2
164  RM Belo Horizonte              2
165  RM Rio de Janeiro              0
166  RM Rio de Janeiro              0
167  RM Rio de Janeiro              0
168  RM Rio de Janeiro              0
169  RM Rio de Janeiro              0
170       RM São Paulo              0
171       RM São Paulo              0
172       RM São Paulo              0
173       RM São Paulo              0
174       RM São Paulo              0
175        RM Curitiba              2
176        RM Curitiba              2
177        RM Curitiba              2
178        RM Curitiba              2
179        RM Curitiba              2
180    RM Porto Alegre              2
181    RM Porto Alegre              2
182    RM Porto Alegre              2
183    RM Porto Alegre              2
184    RM Porto Alegre              2
185            RIDE DF              2
186            RIDE DF              2
187            RIDE DF              2
188            RIDE DF              2
189            RIDE DF              2

[190 rows x 2 columns]

          NOME_AGREGA  clusterNumber
0               BRASIL              0
1               BRASIL              0
2               BRASIL              0
3               BRASIL              0
4               BRASIL              0
5                   RO              0
6                   RO              0
7                   RO              0
8                   RO              0
9                   RO              0
10                  AC              0
11                  AC              0
12                  AC              0
13                  AC              0
14                  AC              0
15                  AM              0
16                  AM              0
17                  AM              0
18                  AM              0
19                  AM              0
20                  RR              0
21                  RR              0
22                  RR              0
23                  RR              0
24                  RR              0
25                  PA              0
26                  PA              0
27                  PA              0
28                  PA              0
29                  PA              0
30                  AP              0
31                  AP              0
32                  AP              0
33                  AP              0
34                  AP              0
35                  TO              0
36                  TO              0
37                  TO              0
38                  TO              0
39                  TO              0
40                  MA              0
41                  MA              0
42                  MA              0
43                  MA              0
44                  MA              0
45                  PI              0
46                  PI              0
47                  PI              0
48                  PI              0
49                  PI              0
50                  CE              0
51                  CE              0
52                  CE              0
53                  CE              0
54                  CE              0
55                  RN              0
56                  RN              0
57                  RN              0
58                  RN              0
59                  RN              0
60                  PB              0
61                  PB              0
62                  PB              0
63                  PB              0
64                  PB              0
65                  PE              0
66                  PE              0
67                  PE              0
68                  PE              0
69                  PE              0
70                  AL              0
71                  AL              0
72                  AL              0
73                  AL              0
74                  AL              0
75                  SE              0
76                  SE              0
77                  SE              0
78                  SE              0
79                  SE              0
80                  BA              0
81                  BA              0
82                  BA              0
83                  BA              0
84                  BA              0
85                  MG              0
86                  MG              0
87                  MG              0
88                  MG              0
89                  MG              0
90                  ES              0
91                  ES              0
92                  ES              0
93                  ES              0
94                  ES              0
95                  RJ              0
96                  RJ              0
97                  RJ              0
98                  RJ              0
99                  RJ              0
100                 SP              0
101                 SP              0
102                 SP              0
103                 SP              0
104                 SP              0
105                 PR              0
106                 PR              0
107                 PR              0
108                 PR              0
109                 PR              0
110                 SC              0
111                 SC              0
112                 SC              0
113                 SC              0
114                 SC              0
115                 RS              0
116                 RS              0
117                 RS              0
118                 RS              0
119                 RS              0
120                 MS              0
121                 MS              0
122                 MS              0
123                 MS              0
124                 MS              0
125                 MT              0
126                 MT              0
127                 MT              0
128                 MT              0
129                 MT              0
130                 GO              0
131                 GO              0
132                 GO              0
133                 GO              0
134                 GO              0
135                 DF              0
136                 DF              0
137                 DF              0
138                 DF              0
139                 DF              0
140           RM Belém              2
141           RM Belém              2
142           RM Belém              2
143           RM Belém              2
144           RM Belém              2
145       RM Fortaleza              2
146       RM Fortaleza              2
147       RM Fortaleza              2
148       RM Fortaleza              2
149       RM Fortaleza              2
150          RM Recife              3
151          RM Recife              3
152          RM Recife              3
153          RM Recife              3
154          RM Recife              3
155        RM Salvador              3
156        RM Salvador              3
157        RM Salvador              3
158        RM Salvador              3
159        RM Salvador              3
160  RM Belo Horizonte              3
161  RM Belo Horizonte              3
162  RM Belo Horizonte              3
163  RM Belo Horizonte              3
164  RM Belo Horizonte              3
165  RM Rio de Janeiro              3
166  RM Rio de Janeiro              3
167  RM Rio de Janeiro              3
168  RM Rio de Janeiro              3
169  RM Rio de Janeiro              3
170       RM São Paulo              3
171       RM São Paulo              3
172       RM São Paulo              3
173       RM São Paulo              3
174       RM São Paulo              3
175        RM Curitiba              1
176        RM Curitiba              1
177        RM Curitiba              1
178        RM Curitiba              1
179        RM Curitiba              1
180    RM Porto Alegre              1
181    RM Porto Alegre              1
182    RM Porto Alegre              1
183    RM Porto Alegre              1
184    RM Porto Alegre              1
185            RIDE DF              1
186            RIDE DF              1
187            RIDE DF              1
188            RIDE DF              1
189            RIDE DF              1